assets/migrations/postgres/004_new_user_parts_columns.sql

```diff
@@ -0,0 +1,114 @@
+-- +goose NO TRANSACTION
+-- This migration adds 3 more columns to the tuple and changelog tables, and fills past and future data with a trigger.
+-- This migration also adds more indexes to the tuple table. It does so concurrently so that the main table isn't locked
+
+-- +goose Up
+
+ALTER TABLE tuple
+    ADD COLUMN IF NOT EXISTS user_object_type TEXT,
+    ADD COLUMN IF NOT EXISTS user_object_id TEXT,
+    ADD COLUMN IF NOT EXISTS user_relation TEXT;
+
+-- jon becomes (..., jon, ...)
+-- user:jon becomes (user, jon, ...)
+-- team:* becomes (team, *, ...)
+-- group:eng#member becomes (group, eng, member)
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION set_new_columns() RETURNS TRIGGER AS $$
+BEGIN
+        NEW.user_object_type = (CASE
+                                    WHEN position(':' in NEW._user) > 0 THEN split_part(NEW._user, ':', 1)
+                                    ELSE ''
+            END),
+
+        NEW.user_object_id = (CASE
+                                  WHEN position('#' in NEW._user) > 0 THEN split_part(split_part(NEW._user, ':', -1), '#', 1)
+                                  ELSE split_part(NEW._user, ':', -1)
+            END),
+
+        NEW.user_relation = (CASE
+                                 WHEN position('#' in NEW._user) > 0 THEN split_part(NEW._user, '#', -1)
+                                 ELSE ''
+            END);
+END;
+$$ language 'plpgsql';
+-- +goose StatementEnd
+
+UPDATE tuple SET
+     user_object_type = (CASE
+                             WHEN position(':' in _user) > 0 THEN split_part(_user, ':', 1)
+                             ELSE ''
+         END),
+     user_object_id = (CASE
+                           WHEN position('#' in _user) > 0 THEN split_part(split_part(_user, ':', -1), '#', 1)
+                           ELSE split_part(_user, ':', -1)
+         END),
+     user_relation = (CASE
+                          WHEN position('#' in _user) > 0 THEN split_part(_user, '#', -1)
+                          ELSE ''
+         END);
+
+CREATE TRIGGER migrate_user_column
+    BEFORE UPDATE ON tuple
+    FOR EACH ROW
+    EXECUTE FUNCTION set_new_columns();
+
+ALTER TABLE tuple ADD CONSTRAINT unique_tuple UNIQUE(store, object_type, object_id, relation, user_object_type, user_object_id, user_relation);
```

Comment:
This establishes an _Access Exclusive_ lock on the whole table, which blocks all read and write operations to the table until the validation constraints have been checked (thus affecting any Check, ListObjects, Expand, Write, Read, etc..). For larger tables (1 million + rows) this causes high latency on read and write operations for existing OpenFGA deployments.

My recommendation is that we add a `NOT VALID` statement to the constraint check which will reduce the impact of the validation check on all existing rows and will avoid the table access exclusive lock, but any inserts/updates after the constraint is added will have to be validated against the uniqueness constraint. We can follow up with `VALIDATE CONSTRAINT` to make sure nothing violated the contraint in the meantime.

See
https://www.postgresql.org/docs/15/explicit-locking.html
https://www.postgresql.org/docs/15/sql-altertable.html

> Scanning a large table to verify a new foreign key or check constraint can take a long time, and other updates to the table are locked out until the ALTER TABLE ADD CONSTRAINT command is committed. The main purpose of the NOT VALID constraint option is to reduce the impact of adding a constraint on concurrent updates. With NOT VALID, the ADD CONSTRAINT command does not scan the table and can be committed immediately. After that, a VALIDATE CONSTRAINT command can be issued to verify that existing rows satisfy the constraint. The validation step does not need to lock out concurrent updates, since it knows that other transactions will be enforcing the constraint for rows that they insert or update; only pre-existing rows need to be checked. Hence, validation acquires only a SHARE UPDATE EXCLUSIVE lock on the table being altered. (If the constraint is a foreign key then a ROW SHARE lock is also required on the table referenced by the constraint.) In addition to improving concurrency, it can be useful to use NOT VALID and VALIDATE CONSTRAINT in cases where the table is known to contain pre-existing violations. Once the constraint is in place, no new violations can be inserted, and the existing problems can be corrected at leisure until VALIDATE CONSTRAINT finally succeeds.

Reply:
> ERROR: UNIQUE constraints cannot be marked NOT VALID

Reply:
I think I need to get rid of this contraint

Reply:
Removed for Postgres (but still there for Mysql)